src/config-manager.js

```diff
@@ -0,0 +1,356 @@
+ */
+
+/**
+ * @type {DisabledFeatures}
+ */
+const disabledFeatures = {
+	redis: false,
+	email: false,
+	captcha: false,
+	s3: false
+};
+
+const requiredFields = [
+	['http.port', 'PN_ACT_CONFIG_HTTP_PORT', Number],
+	['mongoose.uri', 'PN_ACT_CONFIG_MONGO_URI'],
+	['mongoose.database', 'PN_ACT_CONFIG_MONGO_DB_NAME'],
+	['cdn_base', 'PN_ACT_CONFIG_CDN_BASE']
+];
+
+function configure() {
+	if (process.env.PN_ACT_PREFER_ENV_CONFIG === 'true') {
+		logger.info('Loading config from env');
+
+		config = {
+			http: {
+				port: Number(process.env.PN_ACT_CONFIG_HTTP_PORT)
+			},
+			mongoose: {
+				uri: process.env.PN_ACT_CONFIG_MONGO_URI,
+				database: process.env.PN_ACT_CONFIG_MONGO_DB_NAME,
+				options: Object.keys(process.env)
+					.filter(key => key.startsWith('PN_ACT_CONFIG_MONGOOSE_OPTION_'))
+					.reduce((obj, key) => {
+						obj[key.split('_').pop()] = process.env[key];
+						return obj;
+					}, {})
+			},
+			redis: {
+				client: {
+					url: process.env.PN_ACT_CONFIG_REDIS_URL
+				}
+			},
+			email: {
+				host: process.env.PN_ACT_CONFIG_EMAIL_HOST,
+				port: Number(process.env.PN_ACT_CONFIG_EMAIL_PORT),
+				secure: Boolean(process.env.PN_ACT_CONFIG_EMAIL_SECURE),
+				auth: {
+					user: process.env.PN_ACT_CONFIG_EMAIL_USERNAME,
+					pass: process.env.PN_ACT_CONFIG_EMAIL_PASSWORD
+				},
+				from: process.env.PN_ACT_CONFIG_EMAIL_FROM
+			},
+			s3: {
+				endpoint: process.env.PN_ACT_CONFIG_S3_ENDPOINT,
+				key: process.env.PN_ACT_CONFIG_S3_ACCESS_KEY,
+				secret: process.env.PN_ACT_CONFIG_S3_ACCESS_SECRET
+			},
+			hcaptcha: {
+				secret: process.env.PN_ACT_CONFIG_HCAPTCHA_SECRET
+			},
+			cdn_base: process.env.PN_ACT_CONFIG_CDN_BASE,
+			website_base: process.env.PN_ACT_CONFIG_WEBSITE_BASE
+		};
+	} else {
+		logger.info('Loading config from config.json');
+
+		if (!fs.pathExistsSync(`${__dirname}/../config.json`)) {
+			logger.error('Failed to locate config.json file');
+			process.exit(0);
+		}
+
+		config = require(`${__dirname}/../config.json`);
+	}
+
+	logger.info('Config loaded, checking integrity');
+
+	// * Check for required settings
+	for (const requiredField of requiredFields) {
+		const [keyPath, env, convertType] = requiredField;
+
+		const configValue = get(config, keyPath);
+		const envValue = get(process.env, keyPath);
+
+		if (!configValue || (typeof configValue === 'string' && configValue.trim() === '')) {
+			if (!envValue || envValue.trim() === '') {
+				logger.error(`Failed to locate required field ${keyPath}. Set ${keyPath} in config.json or the ${env} environment variable`);
+
+				process.exit(0);
+			} else {
+				logger.info(`${keyPath} not found in config, using environment variable ${env}`);
+
+				const newValue = envValue;
+
+				set(config, keyPath, convertType ? convertType(newValue) : newValue);
+			}
+		}
+	}
+
+	// * Check for optional settings
+
+	const redisConfigValue = get(config, 'redis.client.url');
+	const redisEnvValue = get(process.env, 'PN_ACT_CONFIG_REDIS_URL');
```

Comment:
The flow here feels a bit weird for the `PN_ACT_PREFER_ENV_CONFIG` case, since we already read in that same variable before, thus `redisConfigValue` and `redisEnvValue` are identical.
Are we sure we want mixed env/config file options in the first place? Server usecase would be config.json, Docker usecase would be envvars...

Assuming we want it, maybe some function like..
```js
let configJson = ...;

function getValue(name) {
    if (!PN_ACT_PREFER_ENV_CONFIG && configJson...) {
        return configJson...;
    } else {
        return process.env...;
    }
}

const redis = getValue('redis.client.url');
if (!redis || redis.trim === '') {
    ...
}
set(config, ...);
```
You'd need a way to map between the dot form and envvar names, though `redis.client.url` -> `REDIS_CLIENT_URL` seems like a pretty straightforward transform.

Reply:
> The flow here feels a bit weird for the `PN_ACT_PREFER_ENV_CONFIG` case

I agree it's a bit redundant in this case. The later checks are done just as fail-safes for the optional settings ignoring which option the user preferred. The idea behind it is "fill out the config as best it can depending on how the user wants to load it, then try to fill in any blanks"

> Are we sure we want mixed env/config file options in the first place?

I would say yes. Giving users the option to do so feels better to me, especially considering it's entirely optional. I was also told that having configs be env variables is better for docker, which I plan to run the entire network off docker eventually too

> Assuming we want it, maybe some function like

That's not a bad setup, can definitely work something like that out

Reply:
See https://github.com/PretendoNetwork/account/pull/51#issuecomment-1272691244